api/envoy/config/filter/http/abac/v2alpha/abac.proto

```diff
@@ -0,0 +1,18 @@
+syntax = "proto3";
+
+package envoy.config.filter.http.abac.v2alpha;
+
+option java_outer_classname = "AbacProto";
+option java_multiple_files = true;
+option java_package = "io.envoyproxy.envoy.config.filter.http.abac.v2alpha";
+option go_package = "v2alpha";
+
+import "google/api/expr/v1alpha1/syntax.proto";
+
+// [#protodoc-title: ABAC]
+// Attribute-Based Access Control.
+
+// ABAC filter config.
+message ABAC {
```

Comment:
Instead of adding another filter for access control, can we evolve the RBAC filter as the unified (local) access control filter in Envoy?

The `expr` could just be added to the [Policy](https://github.com/envoyproxy/envoy/blob/27b41b888d5ba624f2a6098c8b9f395e8ef95cbf/api/envoy/config/rbac/v2/rbac.proto#L85) in RBAC, like the following:
```protobuf
message Policy {
  repeated Permission permissions = 1 [(validate.rules).repeated .min_items = 1];

  repeated Principal principals = 2 [(validate.rules).repeated .min_items = 1];

  // The additional conditions that should be satisfied for the request.
  repeated google.api.expr.v1alpha1.Expr conditions = 3;
}
```

This way, it could be easily used together with the other `permission`/`principals` to have more flexible in access control. 

They can share the same support for metric/debugging/experimental already existing in RBAC, for example, the RBAC filter supports shadow policy to experiment a policy without enforcing it. The CEL expression could have the same feature for free if it's added to here. 
The RBAC filter already has the network filter, you will only need minimal code change to support the CEL for non-HTTP requests. And it will also be easier to reuse all the existing e2e tests in not only Envoy code base but also in control plane (e.g. Istio) code base as we can just modifying the policy to add more tests.

A single unified access control filter also makes it much easier for people to learn and adopt, and also gives us a potential an easier path to migrate the old policy to the use CEL-based policy if needed.

Again, I also doesn't think it makes any sense to have the `RBAC`/`ABAC` concept in the data plane level, I have mentioned this in https://github.com/envoyproxy/envoy/issues/6751 together with integrating CEL in RBAC filter. That's just naming if we only need to deal with the current RBAC filter, we can always change the `RBAC` filter to something like `AccessControl` filter later. But it would make it harder to do so if we have both 'ABAC' and 'RBAC' filter.

@lizan @mattklein123 @liminw Would love to hear what do you think about this proposal, thank you!

Reply:
My concern is that RBAC has graduated to a stable filter, which means we can't break API compatibility. I suggest keeping ABAC as alpha for now (even though CEL-CPP is used in production). I plan to refactor bits here to make re-use easier across filters, and we can do something similar to RBAC so that shadow-ing can be re-used.

Reply:
I think the [Breaking change policy](https://github.com/envoyproxy/envoy/blob/master/CONTRIBUTING.md#breaking-change-policy) is mostly about the wire compatibility, I'm not sure how adding a new field could break API compatibility, I think the guideline even allows deprecating fields in v2 if necessary.

And thinking in the big picture, I just feel it's not good to have 2 functional (access control) equivalent filters with the only difference in its DSL. For example, someone may don't like the `ABAC`/`RBAC` API and want to add their `XBAC`, what should we say in this case?

Reply:
I agree with @yangminzhu that if we can transparently add CEL functionality to the existing RBAC filter and not require it, that would be optimal. The only downside I see is that now users of RBAC need to compile/include the CEL runtime. This seems OK, especially if the code is not used unless explicitly configured.

Reply:
Is there a concern that old proxies may not respect CEL conditions even though wire compatibility is preserved? If a condition says `false` then the old proxies will allow traffic.

Reply:
+1 to @yangminzhu and @mattklein123 . We should have only one AccessControl filter, to avoid duplicating all the existing work and confusing our users. What we want to do here is to improve the existing RBAC filter to support CEL functionality, instead of adding another filter for access control.

Reply:
> Is there a concern that old proxies may not respect CEL conditions even though wire compatibility is preserved? If a condition says false then the old proxies will allow traffic.

This is a problem for any added field/condition, not just CEL, so we need to think about this problem holistically anyway.

Reply:
Updated to add a `condition` field to RBAC filter instead.

Reply:
Resolving this since the condition is merged into RBAC Policy.